sdk/src/main/java/io/opentdf/platform/sdk/Autoconfigure.java

```diff
@@ -0,0 +1,749 @@
+    
+            for (SingleAttributeClause clause : e.must) {
+                List<PublicKeyInfo> kcv = new ArrayList<>(clause.values.size());
+    
+                for (AttributeValueFQN term : clause.values) {
+                    KeyAccessGrant grant = byAttribute(term);
+                    if (grant == null) {
+                        throw new AutoConfigureException(String.format("no definition or grant found for [%s]", term));
+                    }
+    
+                    List<String> kases = grant.kases;
+                    if (kases.isEmpty()) {
+                        kases = List.of(RuleTypes.EMPTY_TERM);
+                    }
+    
+                    for (String kas : kases) {
+                        kcv.add(new PublicKeyInfo(kas));
+                    }
+                }
+    
+                String op = ruleToOperator(clause.def.getRule());
+                if (op == RuleTypes.UNSPECIFIED) {
+                    logger.warn("Unknown attribute rule type: " + clause);
+                }
+    
+                KeyClause kc = new KeyClause(op, kcv);
+                kcs.add(kc);
+            }
+    
+            return new BooleanKeyExpression(kcs);
+        }
+
+        AttributeBooleanExpression constructAttributeBoolean() throws AutoConfigureException {
+            Map<String, SingleAttributeClause> prefixes = new HashMap<>();
+            List<String> sortedPrefixes = new ArrayList<>();
+            for (AttributeValueFQN aP : policy) {
+                AttributeNameFQN a = aP.prefix();
+                String p = a.toString().toLowerCase();
+                SingleAttributeClause clause = prefixes.get(p);
+                if (clause != null) {
+                    clause.values.add(aP);
+                } else if (byAttribute(aP) != null) {
+                    var x = new SingleAttributeClause(byAttribute(aP).attr, new ArrayList<AttributeValueFQN>(Arrays.asList(aP)));
+                    prefixes.put(p, x);
+                    sortedPrefixes.add(p);
+                }
+            }
+
+            List<SingleAttributeClause> must = sortedPrefixes.stream()
+                    .map(prefixes::get)
+                    .collect(Collectors.toList());
+
+            return new AttributeBooleanExpression(must);
+        }
+
+
+
+        public class AttributeMapping {
+
+            private Map<AttributeNameFQN, Attribute> dict;
+        
+            public AttributeMapping() {
+                this.dict = new HashMap<>();
+            }
+        
+            public void put(Attribute ad) throws Exception {
+                if (this.dict == null) {
+                    this.dict = new HashMap<>();
+                }
+        
+                AttributeNameFQN prefix;
+                try {
+                    prefix = new AttributeNameFQN(ad.getFqn());
+                } catch (Exception e) {
+                    throw new Exception(e);
+                }
```

Comment:
an AutoConfigureException here indicates an invalid attribute FQN in the database

```suggestion
            public void put(Attribute ad) throws AutoConfigureException {
                if (this.dict == null) {
                    this.dict = new HashMap<>();
                }
        
                AttributeNameFQN prefix;
                try {
                    prefix = new AttributeNameFQN(ad.getFqn());
                } catch (AutoConfigureException e) {
                    throw new AutoConfigureException(e);
                }
```